### PR TITLE
API: Add ``NumpyUnpickler``

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @ngoldbaum,

Connected to #24634.

This is a separate PR with `NumpyUnpickler` that is meant to load 1.x array pickles with numpy 2.0. 
(Without it a warning would be raised, and once `numpy.core` is gone - an exception).

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Can this get the docs additions from the other PR and a separate release note?

**Comment by mtsokol** [comment]
There's github incident that impacts this PR: https://www.githubstatus.com/incidents/9fk4c2w43clz, so we need to wait.

File: numpy/lib/format.py
Comment (BvB93) [inline numpy/lib/format.py:1]: Considering `NumpyUnpickler` is a new public member of the module, it should in principle be added to `__all__`.

Comment (mtsokol) [reply numpy/lib/format.py:1]: Sure! I added it to `__all__`.

File: numpy/lib/format.py [hunk 1]
```diff
@@ -1011,3 +1011,27 @@ def isfileobj(f):
...
+            module.startswith("numpy.core") and
+            numpy_major_version >= 2 and
+            hasattr(np, "_core")
+        ):
+            module = module.replace("core", "_core", 1)
+        elif (
+            module.startswith("numpy._core") and
+            numpy_major_version <= 1
+        ):
+            module = module.replace("_core", "core", 1)
```
Comment (ngoldbaum) [inline numpy/lib/format.py:1036]: This is a little hard to read right now and is also doing unnecessary work. You only need to see if the path starts with `np.core` once. Also it's a little weird to have numpy version checks inside numpy. @charris do you think @mtsokol should do a separate PR manually backporting this with a slightly different implementation?

File: numpy/lib/_npyio_impl.py [hunk 2]
```diff
@@ -459,7 +459,7 @@ def load(file, mmap_mode=None, allow_pickle=False, fix_imports=True,
                 raise ValueError("Cannot load file containing pickled data "
                                  "when allow_pickle=False")
             try:
-                return pickle.load(fid, **pickle_kwargs)
+                return format.NumpyUnpickler(fid, **pickle_kwargs).load()
```
Comment (ngoldbaum) [inline numpy/lib/_npyio_impl.py:462]: Should we also wrap the call to `pickle.dump` to replace `_core` with `core` so numpy 2.x doesn't write `npy` files that are unreadable by numpy 1.x. Leave a comment that in a few years when downstream (hopefully) no longer cares about supporting NumPy 1.x we can remove that.

I don't think `NumpyUnpickler` should do this, it should only be done in the numpy file format internals to make extra sure files written in the for at are as compatible as possible.

**Review by ngoldbaum — COMMENTED** [review]
Thanks for moving this forward! Just a few more comments. Let me know if you disagree about wrapping `pickle.dumps` for writing npy files.

File: numpy/lib/format.py (hunk 1 above)
Comment (mtsokol) [reply numpy/lib/format.py:1036]: I understand it this way: We would like a `NumpyUnpickler` that is able to: 
1. Read 1.x pickles when numpy 2.0 is present
2. Read 2.0 pickles when numpy 1.x is present

In this implementation, if the pickle contains `numpy.core` and numpy 2.0 is installed, we need to replace `core` with `_core` (thus we need to check numpy version). 
If numpy 1.x is present then the path should be left untouched.
If pickle contains `numpy._core` and numpy 1.x is present then we need to change `_core` to `core` so that pickle can be loaded.

I think importing numpy is important here to be aware of the installed version and make path updates based on that.

Comment (ngoldbaum) [reply numpy/lib/format.py:1036]: We're inside the numpy source code so we know what version of numpy we're writing code for *a priori* - no need to do the check at runtime. Of course when you backport this code the check will be different, that's why I suggested manually backporting it. Does that make sense?

File: numpy/lib/_npyio_impl.py (hunk 2 above)
Comment (mtsokol) [reply numpy/lib/_npyio_impl.py:462]: I think that each version (1.x and 2.0) should make their own pickles (with `core` and `_core` respectively). And user can employ `NumpyUnpickler` to read cross-version pickles.

If we replace `pickle.dump` in functions in `npyio` then it would mean that `np.save` creates numpy 1.x readable files but direct call to `pickle.dumps(arr)` won't - I think it would be confusing. I think that if someone is operating in numpy 2.0 realm only, they should be able to create pickles any possible way (with `np.save` and `pickle`) and read them without `NumpyUnpickler`.

File: numpy/lib/format.py (hunk 1 above)
Comment (mtsokol) [reply numpy/lib/format.py:1036]: Ah, now I get it. Then I can make separate implementation for 1.x (This implementation is version agnostic - I think two simpler implementations for 1.x and 2.0 are fine for me).

File: numpy/lib/_npyio_impl.py (hunk 2 above)
Comment (ngoldbaum) [reply numpy/lib/_npyio_impl.py:462]: You convinced me. It would be weird for numpy to generate two different kinds of pickle data depending on how you do it. Point withdrawn.

**Comment by charris** [comment]
> do you think @mtsokol should do a separate PR

It is pretty small, but separate would be a bit cleaner. I'm not too concerned about it either way.

Has it been decided that this is better than having both `core` and `_core` in 1.26 and 2.0?

**Comment by ngoldbaum** [comment]
> Has it been decided that this is better than having both core and _core in 1.26 and 2.0?

We'd like to see if we can do it this way. If it turns out to be problematic we'll revert it and go with leaving behind a stub `core` module. We'd like to avoid having the stub module if we can since it'll be pretty confusing and might be fragile because we have no idea what people have stored in pickle files that might need to be whitelisted.

File: doc/source/user/how-to-io.rst [hunk 3]
```diff
@@ -319,6 +319,11 @@ Use :func:`numpy.save` and :func:`numpy.load`.  Set ``allow_pickle=False``,
 unless the array dtype includes Python objects, in which case pickling is
 required.
 
+:func:`numpy.load` also supports unpickling files created with NumPy 1.x.
+If you try to unpickle a 1.x pickled array with an object dtype directly, you
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply doc/source/user/how-to-io.rst]: Sure! Updated.

**Comment by ngoldbaum** [comment]
So far I haven't heard any objections or concrete alternative suggestions for the issue the new API addition solves, so I'm going to merge this to unblock the `core` to `_core` refactor. Thanks @mtsokol!
